### ISSUE TITLE
--ext test.js not working with multi-dot notation

### ISSUE THREAD

**Issue by kentcdodds** [eslint/eslint#2862]
Here's a reproducible example:

```
~/Desktop
😄  $ echo "console.log('hi mom');" > temp.test.js
~/Desktop
😄  $ eslint . --ext test.js
~/Desktop
😄  $ eslint .

temp.test.js
  1:0   error  Unexpected console statement  no-console
  1:0   error  "console" is not defined      no-undef
  1:12  error  Strings must use doublequote  quotes

✖ 3 problems (3 errors, 0 warnings)

~/Desktop
😄  $ eslint . --ext .test.js
~/Desktop
😄  $ eslint . --ext js
~/Desktop
😄  $ eslint . --ext .js

temp.test.js
  1:0   error  Unexpected console statement  no-console
  1:0   error  "console" is not defined      no-undef
  1:12  error  Strings must use doublequote  quotes

✖ 3 problems (3 errors, 0 warnings)
```

It's arguable whether this is expected behavior, but it would certainly be nice if the `--ext` worked more like "files ending in x" rather than "matching x from the last dot to the end of the filename"

On this note, I asked/suggested something [on Gitter](https://gitter.im/eslint/eslint?at=558e7998666fd9af67366bd2) a moment ago that I think would solve my true use-case for something like this:

> Hey friends, is it possible to scope eslint to scope eslint config to files matching a naming convention? For example, I want all of my `*.test.js` files to have the `env`of `mocha` in addition to `browser`.
> I would actually really like this for a bunch of things because I have different rules between my tests and my code (max-len for longer `it` descriptions, or max-nested-callbacks for nesting `describe`s with `it`s). Would be very cool if the `.eslintrc` allowed me to specify a subset of overrides that apply only to my test files. I suppose I could accomplish this with webpack loaders........ hmmmm... yeah :-)
> But not having to manage that myself and being able to include it right in my `.eslintrc` would be magical. Has there been discussion about this before?
> Oh, and by the way, I mix my test files with my source files, so I can't simply plop an `.eslintrc` into my test directory because that doesn't exist. I think this is fairly common.

@bebraw suggested the `--ext` option which lead to discovering and reporting this bug/feature request. But I would really like something like this:

```javascript
{
  // common config goes up here
  "ext": [
    {
      "test": "\\.(test|mock)\\.js$", // regex matching anything ending in .test.js or .mock.js
      "config": {
        // config overrides for test/mock files here
      }
    },
    {
      "test": "^((?!\\.(test|mock)\\.).)*\\.js$", // regex matching anything not including .test. or .mock.
      "config": {
        // config overrides for non-test/mock files here
      }
    }
  ]
}
```

This would be fantastically flexible and support a great deal of use cases and conventions without sacrificing simplicity (in my mind anyway). Feedback welcome!

**Comment by kentcdodds** [comment]
Note, in this case, perhaps `ext` would not be an appropriate config property name because it allows you to specify arbitrary regex to match the filename to. Something like `filenameOverrides` may be more appropriate.

**Comment by michaelficarra** [comment]
Would replacing extension with suffix not be sufficient?

**Comment by ilyavolodin** [comment]
@kentcdodds The feature you are asking for has been proposed for v2.0.0 release. Being able to specify patterns and assign configs to them will be done after we release v1.0.0. `ext` was never meant for this. Currently eslint will only process `.js` files, `ext` is there to allow users to process other file types (like `.jsx`, for example).

**Comment by nzakas** [comment]
--ext is only the extension, not a suffix, so this is behaving correctly. See: http://eslint.org/docs/user-guide/command-line-interface#ext

As Ilya mentioned, we are discussing something like this for post 1.0.0, but it's not a sure thing as it's quite complex. So, closing this issue as it's not currently on the roadmap. You can follow #2203 for updates.
